rj_param_utils/include/rj_param_utils/param.h

```diff
@@ -0,0 +1,167 @@
+ * happens in the constructor.
+ */
+class ParamRegisterer {
+public:
+    template <typename ParamType>
+    ParamRegisterer(const char* name, const char* help, const char* filename,
+                    ParamType& current_storage);
+};
+}  // namespace internal
+
+/**
+ * @brief A Param represents a parameter.
+ * @tparam T The type of the parameter.
+ */
+template <typename T>
+class Param {
+public:
+    using Ptr = std::unique_ptr<Param>;
+
+    Param(const char* name, const char* help, const char* filename, T& param)
+        : name_{name},
+          help_{help},
+          filename_{filename},
+          default_value_{param},
+          param_{param} {}
+
+    /**
+     * @brief Updates the parameter with the new value.
+     * @param new_value The new value of the parameter.
+     * @throws std::runtime_error if the type is set.
+     */
+    void Update(T&& new_value) { param_ = std::move(new_value); }
+    void Update(const T& new_value) { param_ = new_value; }
+
+    const T& value() const { return param_; }
+
+    const T& default_value() const { return default_value_; }
+
+    [[nodiscard]] const std::string& name() const { return name_; }
+
+    [[nodiscard]] const std::string& help() const { return help_; }
+
+    template <typename ParamType>
+    friend std::ostream& operator<<(std::ostream& os,
+                                    const Param<ParamType>& param);
+
+private:
+    const std::string name_;
+    const std::string help_;
+    const std::string filename_;
+    const T& default_value_;
+    T& param_;
+};
+
+/**
+ * @brief A ParamProvider represents a "provider" for parameters. In particular,
+ * it should provide a way of updating the parameters.
+ */
+class ParamProvider {
+public:
+    template <typename T>
+    using ParamMap = std::unordered_map<std::string, typename Param<T>::Ptr>;
+
+    template <typename ParamType>
+    bool Get(const std::string& param_name, ParamType* value) const;
+
+    template <typename ParamType>
+    void Update(const std::string& param_name, const ParamType& new_value);
+
+    template <typename ParamType>
+    bool TryUpdate(const std::string& param_name, const ParamType& new_value);
+
+    template <typename ParamType>
+    ParamMap<ParamType>& GetParamMap();
+};
+
+}  // namespace params
+
+/**
+ * @brief Defines a namespaced parameter. The defined parameter can be used as
+ * PARAM_#, where * denotes # denotes the name passed in.
+ * @param type The type of the parameter.
+ * @param prefix The prefix / namespace of the parameter, using . as a
+ * separator.
+ * @param name The variable name of the parameter.
+ * @param val The default value of the parameter.
+ * @param description A description of the parameter.
+ */
+#define DEFINE_NAMESPACED_VARIABLE(type, prefix, name, val, description)   \
+    namespace params::variables {                                          \
+    static type PARAM_##name##_storage = val;                              \
```

Comment:
Yeah, in that case it will conflict. I don't think it will conflict if they're in different translation units though.

Do you think this use case should be considered?

Reply:
PID constants for different controllers is a good example. XY and Heading controllers both use a PID controller but with different gains

Reply:
https://github.com/RoboJackets/robocup-software/blob/staging/soccer/config/sim.xml#L88
https://github.com/RoboJackets/robocup-software/blob/staging/soccer/config/sim.xml#L97

Reply:
I guess the easiest solution I can think of right now would be to put the `DEFINE_XXX` inside different namespaces.

Reply:
Is it not possible to attach the prefix to all the stuff here? So like `PARAM_##prefix##__##name##_storage` or something like that. The prefix + name should be unique imo

Reply:
The problem with prefix is that ROS2 uses `.` as the separator for namespaces, ie. `a.b.c.example_param`, and `.` is an invalid character for C++ identifiers.

One way I could see this working would be to use `::` as the separator for the prefixes, ie.
`DEFINE_NAMESPACED_BOOL(a::b::c, thing, true, "description.")`, and then access it as `a::b::c::PARAM_thing` or something, and then do the conversion between `::` and `.` in `ROS2ParamProvider`.